jmclient/jmclient/payjoin.py

```diff
@@ -0,0 +1,503 @@
+            if inp.derivation_map:
+                return False
+        for out in self.initial_psbt.outputs:
+            if out.derivation_map:
+                return False
+
+        # TODO we can replicate the mempool check here for
+        # Core versions sufficiently high, also encapsulate
+        # it in bc_interface.
+
+        # our logic requires no more than one change output
+        # for now:
+        found_payment = 0
+        assert len(self.payment_tx.vout) in [1, 2]
+        for out in self.payment_tx.vout:
+            if out.nValue == self.amount and \
+               btc.CCoinAddress.from_scriptPubKey(
+                   out.scriptPubKey) == self.destination:
+                found_payment += 1
+            else:
+                # store this for our balance check
+                # for receiver proposal
+                self.change_out = out
+        if not found_payment == 1:
+            return False
+
+        return True
+
+    def check_receiver_proposal(self, in_pbst, signed_psbt_for_fees):
+        """ This is the most security critical part of the
+        business logic of the payjoin. We must check in detail
+        that what the server proposes does not unfairly take money
+        from us, and also conforms to acceptable structure.
+        We perform the following checks of the receiver proposal:
+         1. Check that there are more inputs (i.e. some were contributed).
+         2. Does it contain our inputs, unchanged?
+         3. Does it contain our payment output, with amount increased?
+         4. Are the other inputs finalized, and of the correct type?
+         5. Is the feerate unchanged within tolerance?
+         6. Does it contain no xpub information or derivation information?
+         7. Are the sequence numbers unchanged (and all the same) for the inputs?
+         8. Is the nLockTime and version unchanged?
+         9. Is the extra fee we pay in reduced change output less than a doubling?
+
+        If all the above checks pass we will consider this valid, and cosign.
+        Returns:
+        (False, "reason for failure")
+        (True, None)
+        """
+        assert isinstance(in_pbst, btc.PartiallySignedTransaction)
+        orig_psbt = self.initial_psbt
+        assert isinstance(orig_psbt, btc.PartiallySignedTransaction)
+        # 1
+        if len(in_pbst.inputs) <= len(orig_psbt.inputs):
+            return (False, "Receiver did not contribute inputs to payjoin.")
+        # 2
+        ourins = [(i.prevout.hash, i.prevout.n) for i in orig_psbt.unsigned_tx.vin]
+        found = [0] * len(ourins)
+        receiver_input_indices = []
+        for i, inp in enumerate(in_pbst.unsigned_tx.vin):
+            for j, inp2 in enumerate(ourins):
+                if (inp.prevout.hash, inp.prevout.n) == inp2:
+                    found[j] += 1
+                else:
+                    receiver_input_indices.append(i)
+
+        if any([found[i] != 1 for i in range(len(found))]):
```

Comment:
looks heavy, `any([f != 1 for f in found])` should be equivalent

Reply:
Good spot, thanks. Addressed in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/536/commits/9933408ad70e717014ecbb24796d104f3263ac05